### PR TITLE
40351: Term count scaling, approach two

### PR DESCRIPTION
**Pull request by peterwilsoncc**
Completely revised approach to scaling term counting.

* respects deferring of term counting
* replaces calls to the current functions with calls to new functions
* does not introduce new full count points
* terms added in `wp_insert_post()` can be double counted (this was happening with the earlier approach too)
* has a single unit test to demonstrate the bug & prevent passing.

Trac ticket: https://core.trac.wordpress.org/ticket/40351

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/term/getTerms.php [hunk 1]
```diff
@@ -2635,6 +2635,19 @@ public function test_pad_counts_should_work_when_first_taxonomy_is_nonhierarchic
 		}
 	}
 
+	/**
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/term/getTerms.php:2638]: Why this is broken:

* The factory creates a post with a status of published
* `wp_insert_post()` ultimately calls `wp_set_object_terms()` which increments the count as the post is published
* The count is then incremented again on the call to `_update_term_count_on_transition_post_status()`
* Off by one...

File: src/wp-includes/taxonomy.php [hunk 2]
```diff
@@ -3227,12 +3227,178 @@ function wp_defer_term_counting( $defer = null ) {
...
+ * @param string    $taxonomy     The context of the term.
+ * @param int       $decrement_by By how many the term count is to be decremented. Default 1.
+ * @param bool      $do_deferred  Whether to flush the deferred term counts too. Default false.
+ * @return bool If no terms will return false, and if successful will return true.
+ */
+function wp_decrement_term_count( $tt_ids, $taxonomy, $decrement_by = 1, $do_deferred = false ) {
+	return wp_modify_term_count_by( $tt_ids, $taxonomy, $decrement_by * -1, $do_deferred );
+}
+
+/**
```
Comment (peterwilsoncc) [inline src/wp-includes/taxonomy.php:3277]: Maybe call this function as `wp_modify_term_count_by( null, null, null, true )` when do_deferred set in the existing function for backward compatibility.

If the previous function is called and then this one, the term count will be out (and extra db queries are done for no reason) so deferred term counts might need be stored in a new function.

**Review by peterwilsoncc — COMMENTED** [review]
TODO:

* [x] Prevent off by one error/double counting (more tests needed here too)
* [x] Additional tests
* [ ] Ensure modifying the count is correct for non-post objects (ie, users, categories and probably even links)

**Review by NeilWJames — COMMENTED** [review]
I think that the function definition for update_count_by_callback in class-wp-taxonomy. php function set_props needs to consume $modify_by as well, i.e.
return call_user_func( $args['update_count_callback'], $tt_ids, $taxonomy, $modify_by );

Regards,
Neil

**Comment by peterwilsoncc** [comment]
@NeilWJames Thanks for reviewing.

As `update_count_callback()` is expecting two arguments in current releases of WP, leaving `$modify_by` out of the call is intentional.

If a developer has defined a custom counting callback, calling that will ensure their counts are correct until they update how they are registering their taxonomy.

File: src/wp-includes/class-wp-taxonomy.php [hunk 3]
```diff
@@ -180,6 +180,14 @@ final class WP_Taxonomy {
 	 */
 	public $update_count_callback;
 
+	/**
+	 * Function that will called when the count is modified by an amount.
```
Comment (joemcgill) [inline src/wp-includes/class-wp-taxonomy.php:184]: will [be] called

**Pull request closed without merging**